Installer: rewire spliced specs via RewireTask

nhanford opened the pull request:
This PR allows users to configure explicit splicing replacement of an abstract spec in the concretizer.

```
concretizer:
  splice:
    explicit:
    - target: mpi
      replacement: mpich/abcdef
      transitive: true
```
This config block would mean "for any spec that concretizes to use mpi, splice in `mpich/abcdef` in place of the mpi it would naturally concretize to use. See #20262, #26873, #27919, and #46382 for PRs enabling splicing in the Spec object. This PR will be the first place the splice method is used in a user-facing manner. See https://spack.readthedocs.io/en/latest/spack.html#spack.spec.Spec.splice for more information on splicing.

This will allow users to reuse generic public binaries while splicing in the performant local mpi implementation on their system.

In the config file, the `target` may be any abstract spec. The `replacement` must be a spec that includes an abstract hash `/abcdef`. The `transitive` key is optional, defaulting to `true` if left out.

Two important items to note: 
1. When writing explicit splice config, the user is in charge of ensuring that the replacement specs they use are binary compatible with whatever targets they replace. In practice, this will likely require either specific knowledge of what packages will be installed by the user's workflow, or somewhat more specific abstract "target" specs for splicing, to ensure binary compatibility.
2. Explicit splices can cause the output of the concretizer not to satisfy the input. For example, using the config above and consider a package in a binary cache `hdf5/xyzabc` that depends on mvapich2. Then the command `spack install hdf5/xyzabc` will instead install the result of splicing `mpich/abcdef` into `hdf5/xyzabc` in place of whatever mvapich2 spec it previously depended on. When this occurs, a warning message is printed `Warning: explicit splice configuration has caused the the concretized spec {concrete_spec} not to satisfy the input spec {input_spec}".

Highlighted technical details of implementation:

1. This PR required modifying the installer to have two separate types of Tasks, `RewireTask` and `BuildTask`. Spliced specs are queued as `RewireTask` and standard specs are queued as `BuildTask`. Each spliced spec retains a pointer to its `build_spec` for provenance. If a `RewireTask` is dequeued and the associated `build_spec` is neither available in the install_tree nor from a binary cache, the `RewireTask` is requeued with a new dependency on a BuildTask for the build_spec, and `BuildTask`s are queued for the build spec and its dependencies.
2. Relocation is modified so that a spack binary can be simultaneously installed and rewired. This ensures that installing the `build_spec` is not necessary when splicing from a binary cache.
4. The splicing model is modified to more accurately represent build dependencies -- that is, spliced specs do not have build dependencies, as spliced specs are never built. Their build_specs retain the build dependencies, as they may be built as part of installing the spliced spec.
5. There were vestiges of the compiler bootstrapping logic that were not removed in #46237 because I asked @alalazo to leave them in to avoid making the rebase for this PR harder than it needed to be. Those last remains are removed in this PR.

becker33 commented:
In case this rebase goes bad, the last commit before I attempted to rebase was `4cde3f7e4cc25e09be3907fcb0a27f6c2a10cb5f`.

becker33 commented:
TODO list:
- [ ] docs
- [ ] increase test coverage depending on codecov output
- [ ] ensure tests pass

becker33 commented:
@spackbot filx style

becker33 commented:
@spackbot fix style

greenc-FNAL commented:
*[Adding here based on [2023-09-27 telcon discussion](https://github.com/spack/spack/wiki/Telcon:-2023-09-27)]*

1. We have been using #35975 in our local fork for several months now. If the same behavior was achievable with the functionality of this PR that would be welcome. Namely, that packages in BuildCache built with externals be considered for use _at least_ on systems with compatible externals.
2. We have been hoping for some time that one could have a situation where a set of packages built for a generic microarchitecture (say `x86_64v3`) could be usable on a system where one or more packages should be built locally for the native microarchitecture while reusing downstream generic packages from BuildCache. Hypothetical: a DAG with multiple packages using `numpy`—BuildCache has the `x86_64v3` builds, but we wish to have a compatible native build of `numpy` while still getting all the dependents thereof from BuildCache.

haampie commented:
Can there be a human readable description of what this PR is doing?

becker33 committed:
Squashing the following to create an easier rebase
The hashes are preserved so they are available in git reflog if needed

* bd8bdb82ae - (HEAD -> features/spliced-install, nhanford/features/spliced-install) mark test xfail on windows (10 months ago) <Gregory Becker>
* 800451a4a7 - bugfix: data sharing issue between spliced spec and build_spec (10 months ago) <Gregory Becker>
* cf5f0c8539 - original concretizer: apply splice config (10 months ago) <Gregory Becker>
* dce49c199a - update grep test for new splice mock package (10 months ago) <Gregory Becker>
* 22bfeb5978 - update command completion (10 months ago) <Gregory Becker>
* 84049edbc4 - style (10 months ago) <Gregory Becker>
* a4afe8a13c - bugfix: checking virtuals for splices across names (10 months ago) <Gregory Becker>
* 6d757f3d41 - doing spackbot's job for it (10 months ago) <Gregory Becker>
* 87c210db26 - fix config mutability bug in splice test (10 months ago) <Gregory Becker>
* d84b50cb46 - update test for error wording change (10 months ago) <Gregory Becker>
* 3f13dc8975 - fixup after rebase (10 months ago) <Gregory Becker>
* b734f47b13 - fix failure tracker for rebase (10 months ago) <Gregory Becker>
* d84f6c3c7a - catch broader exception to catch tty.die SystemExit (10 months ago) <Gregory Becker>
* 200b7a78a5 - fixup rebase issues (10 months ago) <Gregory Becker>
* 65971cd9c3 - debug failing workflow (10 months ago) <Gregory Becker>
* ca23acc2fc - docs arguments typo (10 months ago) <Gregory Becker>
* c8cfd9143f - bugfix node keys for spliced specs (10 months ago) <Gregory Becker>
* e57758b15b - fix style (10 months ago) <Gregory Becker>
* 5d63a0b886 - update type hints and docs BuiltTask -> Task (10 months ago) <Gregory Becker>
* 5f373b813c - undo reorder classes to make rebase easier after rebase (10 months ago) <Gregory Becker>
* e851b6ed6b - fix rewire for virtual splices across names (10 months ago) <Gregory Becker>
* 839c1a5ba7 - fix environment reading/writing for build_spec entries (10 months ago) <Gregory Becker>
* ae601740d1 - hack it together to work for right now (10 months ago) <Gregory Becker>
* a6bc44ff12 - test rewire task and fix bugs (10 months ago) <Gregory Becker>
* 79f895900d - move fixtures from bindist to conftest and rename (10 months ago) <Gregory Becker>
* 398456d3cc - move splice config to separate section (10 months ago) <Gregory Becker>
* 13df264fd1 - checkpointing RewireTask changes (10 months ago) <Nathan Hanford>
* 5cb808c2ec - test for explicit_splices; fix bug in build_spec for splices across virtual names (10 months ago) <Gregory Becker>
* 019f7600b4 - execute splices from config (10 months ago) <Gregory Becker>
* 6d6b541e9c - move splice configs into concretizer section (10 months ago) <Gregory Becker>
* 06f90a14ea - config schema: add splice section (10 months ago) <Gregory Becker>
* 38e372f0bf - checkpointing tests (10 months ago) <Nathan Hanford>
* 193ca7806f - initial implementation allowing relocate_package to do rewiring tasks (10 months ago) <Gregory Becker>
* 6fb67d1063 - finish debug session with nhanford (10 months ago) <Gregory Becker>
* f201155a4a - checkpointing rewiring (10 months ago) <Nathan Hanford>
* 1e33b40573 - test modifications (10 months ago) <Nathan Hanford>
* 7b5aaf8fa7 - need dependents for install (10 months ago) <Nathan Hanford>
* 8d2df89e8e - refactor passes tests (10 months ago) <Nathan Hanford>
* d918a654d7 - checkpoint refactor to rebase and test (10 months ago) <Nathan Hanford>
* 61d805cfd8 - non-working initial refactor checkpoint (10 months ago) <Nathan Hanford>
* fc0a6431be - reorder classes to make rebase easier (10 months ago) <Gregory Becker>

Co-authored-by: Nathan Hanford <hanford1@llnl.gov>

tldahlgren commented on .github/workflows/valid-style.yml:
```diff
@@ -85,5 +85,5 @@ jobs:
           spack debug report
           spack -d bootstrap now --dev
-          spack style -t black
+          spack -d style -t black
```
Why is this change relevant to this PR?

becker33 replied:
I think it's not, but I also think that we want to keep it at least for now. IIRC we were getting really useless error messages because `spack style` was failing in new and interesting ways, and that's why we added it.

tldahlgren commented on lib/spack/spack/binary_distribution.py:
```diff
@@ -693,8 +693,9 @@ def get_buildfile_manifest(spec):
 
 def hashes_to_prefixes(spec):
     """Return a dictionary of hashes to prefixes for a spec and its deps, excluding externals"""
+    # TODO: Get this to return a tuple value (name, prefix) and modify all callers.
```
Is this still a "TODO" item or should the comment be removed?

tldahlgren commented on lib/spack/spack/schema/splice.py:
```diff
@@ -0,0 +1,36 @@
+# Copyright 2013-2023 Lawrence Livermore National Security, LLC and other
```
```suggestion
# Copyright 2013-2024 Lawrence Livermore National Security, LLC and other
```

tldahlgren reviewed (commented):
*Very* preliminary/minor feedback.

tldahlgren commented on var/spack/repos/builtin.mock/packages/splice-vt/package.py:
```diff
@@ -0,0 +1,24 @@
+# Copyright 2013-2023 Lawrence Livermore National Security, LLC and other
```
```suggestion
# Copyright 2013-2024 Lawrence Livermore National Security, LLC and other
```

tldahlgren commented on lib/spack/spack/test/conftest.py:
```diff
@@ -60,8 +60,11 @@
+from spack.main import SpackCommand
 from spack.util.pattern import Bunch
 
+mirror_cmd = SpackCommand("mirror")
```
Isn't this a "slippery slope" adding commands to test configuration?

tldahlgren commented on lib/spack/spack/test/conftest.py:
```diff
@@ -980,6 +983,38 @@ def install_mockery(temporary_store: spack.store.Store, mutable_config, mock_pac
+
+
+@pytest.fixture(scope="function")
+def temporary_mirror(temporary_mirror_dir):
```
Should this be using `mutable_config`?

tldahlgren commented on lib/spack/spack/test/conftest.py:
```diff
@@ -980,6 +983,38 @@ def install_mockery(temporary_store: spack.store.Store, mutable_config, mock_pac
+@pytest.fixture(scope="function")
+def temporary_mirror(temporary_mirror_dir):
+    mirror_url = url_util.path_to_file_url(temporary_mirror_dir)
+    mirror_cmd("add", "--scope", "site", "test-mirror-func", mirror_url)
+    yield temporary_mirror_dir
+    mirror_cmd("rm", "--scope=site", "test-mirror-func")
```
In addition to adding the import around line 46:

```suggestion
    mirror_name = "test-mirror-func"
    mirror = spack.mirror.Mirror(mirror_url, name=mirror_name)
    spack.mirror.add(mirror, "site")
    yield temporary_mirror_dir
    spack.mirror.remove(mirror_name, "site")
```

🤔 Part of me is wondering if this should be within a config override context manager but I've only run this by one of the unit test so haven't yet convinced myself one way or the other.

tldahlgren commented on lib/spack/spack/test/conftest.py:
```diff
@@ -980,6 +983,38 @@ def install_mockery(temporary_store: spack.store.Store, mutable_config, mock_pac
+@pytest.fixture(scope="function")
+def mutable_temporary_mirror(mutable_temporary_mirror_dir):
+    mirror_url = url_util.path_to_file_url(mutable_temporary_mirror_dir)
+    mirror_cmd("add", "--scope", "site", "test-mirror-func", mirror_url)
+    yield mutable_temporary_mirror_dir
+    mirror_cmd("rm", "--scope=site", "test-mirror-func")
```
Same here for NOT using a command in `conftest.py`.

tldahlgren commented on lib/spack/spack/test/installer.py:
```diff
@@ -740,6 +741,85 @@ def test_installer_init_requests(install_mockery):
+
+    out = original_spec.splice(spec_to_splice, transitive)
+
+    buildcache = SpackCommand("buildcache")
+    buildcache(
+        "push",
+        "--allow-root",
+        "--unsigned",
+        "--update-index",
+        mutable_temporary_mirror,
+        str(original_spec),
+        str(spec_to_splice),
+    )
```
Same issue here with using a `SpackCommand` in a non-command test module.  Unfortunately, not doing so requires a refactor of `cmd/buildcache.py` (though that would be the right thing to do to expose an API for use here).

becker33 committed:
improve install output

becker33 committed:
rewiring: new dependencies may not have relocation information

tldahlgren commented on lib/spack/spack/installer.py:
```diff
@@ -902,13 +932,13 @@ def __init__(
             request: the associated install request where ``None`` can be
                 used to indicate the package was explicitly requested by the user
-            compiler: whether task is for a bootstrap compiler
+            compiler (bool): whether task is for a bootstrap compiler
```
This change is not needed since already specified on line 928.

tldahlgren commented on lib/spack/spack/installer.py:
```diff
@@ -2481,7 +2649,7 @@ def install(self):
         try:
             with fs.replace_directory_transaction(self.task.pkg.prefix):
-                self.installer._install_task(self.task, self.install_status)
+                return self.installer._install_task(self.task, self.install_status)
```
This method does not return a value nor does it appear to need to return at this point.

becker33 committed:
fix rebase artifact

becker33 committed:
remove completed TODO

becker33 committed:
minor review requests

becker33 committed:
test/installer.py: fix rebase artifacts

becker33 committed:
attempt to solve windows bootstrap problem

tldahlgren commented on lib/spack/spack/installer.py:
```diff
@@ -999,6 +998,33 @@ def __init__(
+                dep = spack.compilers.pkg_spec_for_compiler(compiler_spec)
+                dep.constrain("platform=%s" % str(arch_spec.platform))
+                dep.constrain("os=%s" % str(arch_spec.os))
+                dep.constrain("target=%s:" % arch_spec.target.microarchitecture.family.name)
```
Should be converted to f-strings.

tldahlgren commented on lib/spack/spack/installer.py:
```diff
@@ -1052,6 +1078,21 @@ def add_dependent(self, pkg_id: str) -> None:
+            installed (bool):  install status of the dependency package
+        """
+        if pkg_id != self.pkg_id and pkg_id not in self.dependencies:
+            tty.debug("Adding {0} as a depencency of {1}".format(pkg_id, self.pkg_id))
```
Should be f-string.

tldahlgren commented on lib/spack/spack/installer.py:
```diff
@@ -1068,6 +1109,39 @@ def flag_installed(self, installed: List[str]) -> None:
+        # Move to a module level method.
+        if not os.path.exists(pkg.spec.prefix):
+            path = spack.util.path.debug_padded_filter(pkg.spec.prefix)
+            tty.debug("Creating the installation directory {0}".format(path))
```
Should be f-string.

tldahlgren commented on lib/spack/spack/installer.py:
```diff
@@ -1112,6 +1186,99 @@ def priority(self):
+            # stop early from clients, and is not an error at this point
+            pid = "{0}: ".format(self.pid) if tty.show_pid() else ""
+            tty.debug("{0}{1}".format(pid, str(e)))
+            tty.debug("Package stage directory: {0}".format(pkg.stage.source_path))
```
Should be f-strings.

tldahlgren commented on lib/spack/spack/installer.py:
```diff
@@ -1221,7 +1391,7 @@ def _add_init_task(
     ) -> None:
         """
-        Creates and queus the initial build task for the package.
+        Creates and queus the initial task for the package.
```
Typo: `queus` -> `queues`.

tldahlgren commented on lib/spack/spack/installer.py:
```diff
@@ -1521,6 +1693,75 @@ def _ensure_locked(
+                    spack.store.STORE.failure_tracker.clear(dep, force=False)
+                    self._add_init_task(dep.package, task.request, False, self.all_dependencies)
+
+            compiler = spec.build_spec.compiler
+            arch = spec.build_spec.architecture
+
+            if compiler not in packages_per_compiler:
+                packages_per_compiler[compiler] = {}
+
+            if arch not in packages_per_compiler[compiler]:
+                packages_per_compiler[compiler][arch] = []
+
+            packages_per_compiler[compiler][arch].append(spec.build_spec.package)
```
What is a use case where this is necessary (i.e., not handled by the traversal loop starting at line 1708)?

becker33 replied:
I think this may be unnecessary, but I will try to confirm tomorrow

tldahlgren committed:
installer.py tweaks:
- convert formtas to f-strings in updated/moved code;
- add type hint (back) to _add_compiler_package_to_config;
- fix typo: queus -> queues
- add initial docstring to _requeue_with_build_spec_tasks

tldahlgren committed:
test_install_task_requeue_build_specs: added for _requeue_with_build_spec_tasks coverage improvement

tldahlgren reviewed (commented):
Nits

tldahlgren committed:
Add test_rewire_virtuals_mpi

tldahlgren committed:
test_rewire_virtuals_mpi: add checks of the provenance of the specs

tldahlgren committed:
Spec.splice: add type hints and rework docstring

tldahlgren committed:
Rebase cleanup: drop test_install_task_add_compiler (#45784), fix typo

tldahlgren commented on lib/spack/spack/binary_distribution.py:
```diff
@@ -2517,6 +2532,10 @@ def _ensure_common_prefix(tar: tarfile.TarFile) -> str:
 
 
 def install_root_node(spec, unsigned=False, force=False, sha256=None):
+    # TODO?: Go through this and determine where it needs to be the deploy_spec vs. the spec.
+    # TODO?: Modify dependency lookup s.t. if it's not there, figure out what virtual it provided
+    # and find out what is now providing that virtual. Else raise an error that it didn't exist.
+    # Inverts the logic where you would call the build_spec as spec and the spec as deploy_spec
```
Does this still need to be addressed?

becker33 replied:
I don't believe so. I think we discovered that the correct behavior fell out naturally once we got `relocate_package` modified

tldahlgren commented on lib/spack/spack/spec.py:
```diff
@@ -4150,77 +4150,103 @@ def trim(self, dep_name):
+        of the spliced spec and any correspondingly modified dependency specs.
+        This, the original, spec's provenance is unchanged.
+
         TODO: Extend this for non-concrete Specs.
```
Is this relevant/appropriate to the new method?

becker33 replied:
No, good catch

tldahlgren commented on lib/spack/spack/spec.py:
```diff
@@ -4150,77 +4150,103 @@ def trim(self, dep_name):
+            }
             # deps_to_replace = [self[v] for v in virtuals_to_replace]
         else:
             # TODO: sanity check and error raise here for other.name not in self
```
Should this be addressed in this PR?

tldahlgren commented on lib/spack/spack/spec.py:
```diff
@@ -4242,7 +4268,7 @@ def from_self(name, transitive):
                 if name == other.name:
                     return False
-                if any(
+                if any(  # TODO: should this be all
```
Does this need to be addressed in this PR?

becker33 replied:
No, this is a future TODO. modified it to make it clearer what it's talking about for long-term tracking

tldahlgren commented on lib/spack/spack/test/installer.py:
```diff
@@ -944,79 +1043,76 @@ def test_install_failed_not_fast(install_mockery, monkeypatch, capsys):
-
     installer = create_installer([spec_name], {})
-
+    # TODO: Clean this up in fixture with delattr.
```
Should this be address in this PR?

tldahlgren commented on lib/spack/spack/installer.py:
```diff
@@ -888,9 +906,10 @@ def traverse_dependencies(self, spec=None, visited=None) -> Iterator["spack.spec
+class Task:
+    """Base class for representing a task for a package."""
 
+    # TODO: Consider adding pid as a parameter here:
```
Does this still need to be addressed?

becker33 replied:
Removed

tldahlgren commented on lib/spack/spack/installer.py:
```diff
@@ -938,13 +957,13 @@ def __init__(
-        # Package is associated with a bootstrap compiler
-        self.compiler = compiler
+        # Getting the PID again because it will be needed for execute functionality.
+        # TODO: Should this be cached in PackageInstaller?
```
Here's another `TODO` that needs to be re-evaluated.

becker33 replied:
I don't think it's worth addressing -- any potential time saving by caching is is absolutely trivial compared to the install time. I'll remove the TODO.

tldahlgren commented on lib/spack/spack/installer.py:
```diff
@@ -1112,6 +1186,99 @@ def priority(self):
+    """Class for representing a rewire task for a package."""
+
+    def execute(self, install_status):
+        # TODO: Docstring
```
Need docstring here.

tldahlgren commented on lib/spack/spack/installer.py:
```diff
@@ -1521,6 +1693,75 @@ def _ensure_locked(
+    def _requeue_with_build_spec_tasks(self, task):
+        """Requeue the task and its missing build spec dependencies"""
+        # Full install of the build_spec is necessary because it didn't already exist somewhere
+        # TODO: Bootstrap compilers first (from add_tasks)
```
Still needed?

becker33 replied:
No

tldahlgren commented on lib/spack/spack/installer.py:
```diff
@@ -1521,6 +1693,75 @@ def _ensure_locked(
+
+            for compiler, archs in packages_per_compiler.items():
+                for arch, packages in archs.items():
+                    # TODO: Ensure that this works w.r.t all deps
```
Is this still relevant?

becker33 replied:
I don't believe so.

tldahlgren commented on lib/spack/spack/installer.py:
```diff
@@ -1609,81 +1850,24 @@ def _add_tasks(self, request: BuildRequest, all_deps):
-            pid = f"{self.pid}: " if tty.show_pid() else ""
-            tty.debug(f"{pid}{str(e)}")
-            tty.debug(f"Package stage directory: {pkg.stage.source_path}")
+        # TODO: use install_status
```
What was the intended change here?

becker33 replied:
That TODO is already addressed, I'm removing it

tldahlgren reviewed (changes requested):
Flagged the TODOs to be revisited and comments removed or addressed or deferred.  Some could be clearer if deferred.

becker33 committed:
address review comments

becker33 committed:
remove TODO

tldahlgren committed:
test_rewire_virtual: switch from mpileaks to splice-vt (for windows testing)

tldahlgren committed:
test_rewire_virtual: replace provenance checks with prefix checks

becker33 committed:
installer test root from binary: add case splicing virtual

becker33 committed:
test cmd/env: test to round trip spliced spec

tldahlgren reviewed (dismissed):
New tests pass locally for me as well.  

Will ignore uncovered compiler-related changes per off-line discussion.

haampie commented on lib/spack/spack/installer.py:
```diff
@@ -1112,6 +1184,107 @@ def priority(self):
+            try:
+                install_args = self.request.install_args
+                unsigned = install_args.get("unsigned")
+                binary_distribution.install_root_node(self.pkg.spec, unsigned=unsigned)
```
Why are you using `install_root_node`? It's used for bootstrapping (which should've gone through the installer in the first place instead of reinventing the wheel... but that ship has sailed). 

Previously we had `_process_binary_cache_tarball`, which has various actually useful timers.

becker33 replied:
Oversight, not a particular reason. Trying it out with `_process_binary_cache_tarball` now.

haampie commented on lib/spack/spack/installer.py:
```diff
@@ -1112,6 +1184,107 @@ def priority(self):
+            # Currently this is how RPATH-like behavior is achieved on Windows, after install
+            # establish runtime linkage via Windows Runtime link object
+            # Note: this is a no-op on non Windows platforms
+            pkg.windows_establish_runtime_linkage()
```
This is probably a leftover of a rebase gone wrong. It's moved into a post install hook, it does not belong here.

becker33 committed:
windows post-install hook: remove rebase artifact

becker33 committed:
installer: use _process_binary_cache_tarball for binary rewire

becker33 commented:
@haampie I think I've addressed both of your comments

becker33 committed:
remove duplicate code

becker33 committed:
move explicit splice config under concretizer section

becker33 committed:
update bash completions

becker33 committed:
fix schema missing layer

tgamblin commented on lib/spack/spack/binary_distribution.py:
```diff
@@ -715,7 +715,7 @@ def get_buildfile_manifest(spec):
 def hashes_to_prefixes(spec):
```
I think this function should do less, not more, b/c its real purpose has been obfuscated over time:
1. At some point it was probably useful as a helper to create the value needed in `get_buildinfo_dict`, but it's used in multiple places now.
2. All it's doing is repackaging information that's already on specs, and making them harder to read.
3. The *main* functionality of this, which is the somewhat special traversal (link deps + direct, non-external run deps) isn't even captured in the name and it's just buried in the function.

I would suggest something like:

```python
def relocatable_deps(spec):
    deps = [
        s for s in itertools.chain(
            spec.traverse(root=True, deptype="link"), spec.dependencies(deptype="run")
        )
        if not s.external
    ]
    return llnl.util.lang.dedupe(deps, key=lambda s: s.dag_hash())
```

With a descriptive comment about why these specific deps are important to binary distribution.

becker33 replied:
That is much cleaner, thanks! Addressing all related comments now.

tgamblin commented on lib/spack/spack/binary_distribution.py:
```diff
@@ -715,7 +715,7 @@ def get_buildfile_manifest(spec):
     """Return a dictionary of hashes to prefixes for a spec and its deps, excluding externals"""
     return {
-        s.dag_hash(): str(s.prefix)
+        s.dag_hash(): (s.name, str(s.prefix))
```
This just makes a new data structure that's hard to traverse, when we already have a good one.

tgamblin commented on lib/spack/spack/binary_distribution.py:
```diff
@@ -736,7 +736,7 @@ def get_buildinfo_dict(spec):
         "relocate_links": manifest["link_to_relocate"],
         "hardlinks_deduped": manifest["hardlinks_deduped"],
-        "hash_to_prefix": hashes_to_prefixes(spec),
+        "hash_to_prefix": {h: info[1] for h, info in hashes_to_prefixes(spec).items()},
```
I was originally just going to suggest:
```suggestion
        "hash_to_prefix": {h: prefix for h, (_, prefix) in hashes_to_prefixes(spec).items()},
```

But then the awkward iteration made me wonder if it could be better, so with the refactor above:

```suggestion
        "hash_to_prefix": {dep.dag_hash(): dep.prefix for dep in relocatable_deps(spec)},
```

which I think is way more readable.

tgamblin commented on lib/spack/spack/binary_distribution.py:
```diff
@@ -2197,9 +2196,25 @@ def relocate_package(spec):
+        if lookup_dag_hash in hash_to_old_prefix:
+            old_dep_prefix = hash_to_old_prefix[lookup_dag_hash]
             prefix_to_prefix_bin[old_dep_prefix] = new_dep_prefix
             prefix_to_prefix_text[old_dep_prefix] = new_dep_prefix
```
```suggestion
    for dep in relocatable_deps(spec):
        try:
            lookup_dag_hash = spec.build_spec[dep.name].dag_hash()
        except KeyError:
            dependent_edges = spec[dep.name].edges_from_dependents()
            virtuals = set()
            for edge in dependent_edges:
                virtuals.update(edge.virtuals)
            for virtual in virtuals:
                try:
                    lookup_dag_hash = spec.build_spec[virtual].dag_hash()
                    break
                except KeyError:
                    # This is a new dependency
                    tty.debug(f"{spec} does not have relocation for {dep.name}")

        if lookup_dag_hash in hash_to_old_prefix:
            old_dep_prefix = hash_to_old_prefix[lookup_dag_hash]
            prefix_to_prefix_bin[old_dep_prefix] = dep.prefix
            prefix_to_prefix_text[old_dep_prefix] = dep.prefix
```

tgamblin commented on lib/spack/spack/environment/environment.py:
```diff
@@ -2327,14 +2334,18 @@ def filter_specs(self, reader, json_specs_by_hash, order_concretized):
                 )
 
+            if "build_spec" in node_dict:
+                _, bhash, _ = reader.build_spec_from_node_dict(node_dict)
```
`build_spec_from_node_dict` is poorly named (I was thinking it built a whole spec) but it does not.  Consider renaming at some point.

becker33 replied:
Renaming to `extract_build_spec_info_from_node_dict`, which is wordy but much more self-documenting.

tgamblin commented on lib/spack/spack/installer.py:
```diff
@@ -17,16 +16,18 @@
+Failures to install dependency packages result in removal of their
+dependents' tasks from the current process.  A failure file is also
+written (and locked) so that other processes can detect the failure
+and adjust their tasks accordingly.
```
```suggestion
If a dependency package fails to install, its dependents' tasks will be
removed from the installing process's queue.  A failure file is also written
and locked. Other processes use this file to detect the failure and dequeue
its dependents.
```

tgamblin commented on lib/spack/spack/installer.py:
```diff
@@ -102,7 +104,16 @@ def _write_timer_json(pkg, timer, cache):
+
+class InstallAction(enum.Enum):
     #: Don't perform an install
     NONE = 0
```
Why not make this enum consistent with the one above?

tgamblin commented on lib/spack/spack/installer.py:
```diff
@@ -998,6 +1000,33 @@ def __init__(
+                self.dependencies.add(dep_id)
+
+    def execute(self, install_status):
+        raise NotImplementedError
```
document the `install_status` arg -- what's the contract?

tgamblin commented on lib/spack/spack/installer.py:
```diff
@@ -937,13 +960,12 @@ def __init__(
 
-        # Package is associated with a bootstrap compiler
-        self.compiler = compiler
+        # Getting the PID again because it will be needed for execute functionality.
```
This comment isn't self-contained:
* Why "again"?  Where else?
* "execute functionality": which execute functionality?

```suggestion
        # save the PID, which is used by execute()
```

tgamblin commented on lib/spack/spack/installer.py:
```diff
@@ -998,6 +1000,33 @@ def __init__(
         self.attempts = 0
         self._update()
 
+        # Is this task to install a compiler
```
```suggestion
        # does this task install a compiler
```

tgamblin commented on lib/spack/spack/installer.py:
```diff
@@ -998,6 +1000,33 @@ def __init__(
+
+        # Handle bootstrapped compiler
+        #
+        # The bootstrapped compiler is not a dependency in the spec, but it is
+        # a dependency of the build task. Here we add it to self.dependencies
+        if compiler:
+            compiler_spec = self.pkg.spec.compiler
+            arch_spec = self.pkg.spec.architecture
+            strict = spack.concretize.Concretizer().check_for_compiler_existence
+            if (
+                not spack.compilers.compilers_for_spec(compiler_spec, arch_spec=arch_spec)
+                and not strict
+            ):
+                # The compiler is in the queue, identify it as dependency
+                dep = spack.compilers.pkg_spec_for_compiler(compiler_spec)
+                dep.constrain(f"platform={str(arch_spec.platform)}")
+                dep.constrain(f"os={str(arch_spec.os)}")
+                dep.constrain(f"target={arch_spec.target.microarchitecture.family.name}:")
+                dep.concretize()
+                dep_id = package_id(dep.package.spec)
+                self.dependencies.add(dep_id)
```
We decided to remove bootstrapped compilers -- is this needed anymore?

becker33 replied:
No, it will not be needed if this goes in after the removal of bootstrapping compilers. I think it will be easier if we merge this first though (as the refactoring involved in this is harder to rebase) and then also rip this out as part of that PR. But we should talk through that ordering problem with @alalazo

alalazo replied:
I'm fine with any ordering, but would like to get #46237 in soon (to reduce the size of #45189). I think, if I leave the cleanup of `*Task` attributes to this PR there won't be much overlap. See https://github.com/spack/spack/pull/46237#issuecomment-2333321552

tgamblin commented on lib/spack/spack/installer.py:
```diff
@@ -1017,14 +1046,14 @@ def __ne__(self, other):
 
     def __repr__(self) -> str:
-        """Returns a formal representation of the build task."""
+        """Returns a formal representation of the task."""
```
```suggestion
        """Return a formal representation of the task."""
```

tgamblin commented on lib/spack/spack/installer.py:
```diff
@@ -1017,14 +1046,14 @@ def __ne__(self, other):
 
     def __str__(self) -> str:
-        """Returns a printable version of the build task."""
+        """Returns a printable version of the task."""
```
```suggestion
        """Return a printable version of the task."""
```

tgamblin commented on lib/spack/spack/installer.py:
```diff
@@ -1051,6 +1080,21 @@ def add_dependent(self, pkg_id: str) -> None:
+    def add_dependency(self, pkg_id, installed=False):
+        """
+        Ensure the dependency package id is in the task's list so the task priority will be
+        correct.
```
```suggestion
        """Ensure the package is in this task's `dependencies` list.
```

becker33 committed:
fixup rebase

becker33 committed:
binary_distribution: refactor traversal of transitive link and direct run deps

becker33 committed:
rename ambiguously named method

becker33 committed:
tighten up wording for comments and docstrings

becker33 committed:
installer: consistent use of enum.auto()

becker33 committed:
installer: document contract for Task.execute

becker33 committed:
typing issue

alalazo commented on lib/spack/spack/installer.py:
```diff
@@ -901,7 +924,7 @@ def __init__(
     ):
         """
-        Instantiate a build task for a package.
+        Instantiate a task for a package.
```
This is not necessarily to be done in this PR, but the `attempts` argument is unused. Another change for a following PR that might improve readability is to force using keyword arguments for arguments after `pkg`. This is because, when reading a call:
```python
BuildTask(pkg, request, False, 0, 0, inst.STATUS_ADDED, set())
```
it's kind of hard to understand what's being passed without constantly jumping to the definition.

becker33 replied:
The `attempts` argument is technically used, although it's only to determine whether to print `Queueing` or `Requeueing` in a status update message.

alalazo replied:
Hmm, it seems to me the `Task.attempts` _attribute_ is used but the `__init__` argument is not:

https://github.com/spack/spack/blob/6f5f3d862ccc1e2472c3c77552f626dc13bcd8c6/lib/spack/spack/installer.py#L1000

tldahlgren replied:
@alalazo It is *supposed* to be used.  I have a PR to address this (#46293).  

After discussing with @becker33, I'll make the kwargs change in that same PR.

becker33 replied:
Resolving this as handled in #46293

tgamblin commented on lib/spack/spack/solver/asp.py:
```diff
@@ -516,7 +516,7 @@ def _compute_specs_from_answer_set(self):
             candidate = answer.get(node)
 
-            if candidate and candidate.satisfies(input_spec):
+            if candidate and candidate.build_spec.satisfies(input_spec):
```
Shouldn't both the spec and the build spec satisfy the input spec?  Seems like this at least needs a note (a longer PR description might also help)

tgamblin reviewed (dismissed):
LGTM -- with one remaining comment

haampie commented on lib/spack/spack/binary_distribution.py:
```diff
@@ -2199,11 +2204,26 @@ def relocate_package(spec):
+        try:
+            lookup_dag_hash = spec.build_spec[dep.name].dag_hash()
+        except KeyError:
+            dependent_edges = spec[dep.name].edges_from_dependents()
```
`edges_from_dependents` is almost certainly wrong, it points to nodes that are not dependencies of `spec`.

haampie commented on lib/spack/spack/binary_distribution.py:
```diff
@@ -2199,11 +2204,26 @@ def relocate_package(spec):
-            prefix_to_prefix_text[old_dep_prefix] = new_dep_prefix
+    for dep in deps_to_relocate(spec):
+        try:
+            lookup_dag_hash = spec.build_spec[dep.name].dag_hash()
```
Due to reuse you can have multiple instances of `gcc-runtime` among link deps. So using `spec[name]` is a mistake: the current change would produce a single mapping `<old gcc-runtime prefix> => <new gcc-runtime prefix>` instead of relocating all instances.

becker33 committed:
relocation: only look at dependents within the spec

becker33 committed:
remove redundant name-indexed spec access

becker33 committed:
wip: fix dependency on name matching

becker33 committed:
fixup after rebase

becker33 committed:
installer: remove all compiler bootstrapping

becker33 committed:
update test/buildtask.py to match

becker33 committed:
rewire/bindist: find analogs use set instead of list

haampie commented:
Given e23c56c required keyword args would not hurt

haampie commented on lib/spack/spack/binary_distribution.py:
```diff
@@ -2201,11 +2207,26 @@ def relocate_package(spec):
-            prefix_to_prefix_bin[old_dep_prefix] = new_dep_prefix
-            prefix_to_prefix_text[old_dep_prefix] = new_dep_prefix
+    relocation_specs = deps_to_relocate(spec)
+    build_spec_ids = [id(s) for s in spec.build_spec.traverse(deptype=dt.ALL & ~dt.BUILD)]
```
set instead of list to avoid quadratic complexity?

becker33 replied:
Yes sorry about that, had one commit I hadn't pushed.

becker33 committed:
Spec.splice: ensure spliced specs have no build deps

becker33 commented:
> Given https://github.com/spack/spack/commit/e23c56c90ba45ee1016f6da262c3801fd3485680 required keyword args would not hurt

Tammy implemented that in another PR that I now need to rebase this on.

becker33 committed:
explicit splice: warn on splice that deviates from input spec

becker33 committed:
fixup after rebase

tgamblin commented on lib/spack/spack/binary_distribution.py:
```diff
@@ -712,15 +713,21 @@ def get_buildfile_manifest(spec):
+    This special case of spec traversal is specific to binary relocation. Package relocation is
+    restricted to link and run dependencies. Package relocation needs transitive link dependencies
+    because transitive rpath information appears in libraries/executables, but packages only retain
+    references to their direct run dependencies."""
```
```suggestion
    This is a special spec traversal for dependencies we need to consider when relocating a package.
    
    Package binaries, scripts, and other files may refer to the prefixes of  dependencies, so
    we need to rewrite those locations when dependencies are in a different place at install time
    than they were at build time.
    
    This traversal covers transitive link dependencies and direct run dependencies because:
    1. Spack adds RPATHs for transitive link dependencies so that packages can find needed
       dependency libraries.
    2. Packages may call any of their *direct* run dependencies (and may bake their paths into
       binaries or scripts), so we also need to search for run dependency locations when relocating.
       
    This returns a deduplicated list of transitive link dependencies and direct run dependencies.
    """
```

tgamblin commented on lib/spack/spack/binary_distribution.py:
```diff
@@ -2201,11 +2207,26 @@ def relocate_package(spec):
-            prefix_to_prefix_bin[old_dep_prefix] = new_dep_prefix
-            prefix_to_prefix_text[old_dep_prefix] = new_dep_prefix
+    relocation_specs = deps_to_relocate(spec)
+    build_spec_ids = set(id(s) for s in spec.build_spec.traverse(deptype=dt.ALL & ~dt.BUILD))
```
some sort of comment here to explain why we're looking at the build spec (and what an analog is and how/why it's looked up) is probably needed.

becker33 committed:
documentation: docs for explicit splices

becker33 committed:
allow trivial splices of the root spec

becker33 committed:
deps_to_relocate: improve docstring

Co-authored-by: Todd Gamblin <tgamblin@llnl.gov>

becker33 committed:
add comment on relocation analogs

becker33 committed:
Add missing docs file

alalazo commented on lib/spack/spack/test/concretize.py:
```diff
@@ -2281,6 +2281,30 @@ def test_virtuals_are_annotated_on_edges(self, spec_str):
+
+        spec = spack.spec.Spec("hdf5 ^zmpi").concretized()
+
+        assert spec.satisfies(f"^mpich/{mpich_spec.dag_hash()}")
```
I'll submit a one line fix for this, but I think `mpich` dag hash is not an invariant when `transitive=False`. This test just happen to pass because currently `mpich` in the mock repository has no dependency, so I guess both transitive and non-transitive are behaving the same.

alalazo commented on lib/spack/spack/test/cmd/env.py:
```diff
@@ -803,6 +804,38 @@ def test_user_removed_spec(environment_from_manifest):
+    # The one spec is mpileaks
+    for _, spec in e2.concretized_specs():
+        assert spec.spliced
+        assert spec["mpi"].satisfies(zmpi)
```
I think even this assertion is not an invariant. If the original `zmpi` had a build dependency (in the mock repo it doesn't), then the replacement `zmpi` wouldn't have it - and that changes the hash of the node.
